plutus-core/src/Language/PlutusCore/Evaluation/Machine/Cek.hs

```diff
@@ -376,7 +353,7 @@ computeCek ctx _ (Constant ex val) =
 -- s ; ρ ▻ builtin bn  ↦  s ◅ builtin bn arity arity [] [] ρ
 computeCek ctx env (Builtin ex bn) = do
     -- TODO: budget?
-  arity <- arityOf bn
+  BuiltinRuntime _ arity _ _ <- ask >>= lookupBuiltin bn . cekEnvRuntime
```

Comment:
I've talked with Roman a bit about the idea of doing "HOAS-style" evaluation, where e.g. our "value" for functions is an actual Haskell function. I think that would fit naturally here.

Reply:
> I wonder if it would be feasible to rewrite the AST so that it contains the actual function that you want to execute (along with the cost information) so you didn't have to look it up every time. 

I think it's perfectly feasible. But I'd prefer to do that separately, so that I don't need to fix merge conflicts indefinitely.

> There's obviously a risk that the up-front cost of doing that would exceed any saving in lookup time though.
>
> Oh, but is laziness helping us here?

Not really: `lookupBuiltin` can fail, so `>>=` is strict for it, so monadically preprocessing the AST will require traversing the original AST before the result can be evaluated. Unless we throw an actual exception, which may or may not be sensible.